### PR TITLE
fix: guard against null identifier nodes

### LINKED ISSUE #305: [v4] Crash `Cannot read property 'name' of null` in detect-testing-library-utils getImportedUtilSpecifier
One of our tests has some code that looks like:

```ts
(window as any).__THING = false;
```

This is causing a crash with `Cannot read property 'name' of null`.

Looking at [`getImportedUtilSpecifier`](https://github.com/testing-library/eslint-plugin-testing-library/blob/82ace4946fca958698adcc367ba3e8fa9ab070a4/lib/detect-testing-library-utils.ts#L604-L611), it calls `getPropertyIdentifierNode` and checks `.name` on it, even though that [can return `null`](https://github.com/testing-library/eslint-plugin-testing-library/blob/82ace4946fca958698adcc367ba3e8fa9ab070a4/lib/node-utils.ts#L401). It looks like the root cause might be `isExpressionStatement` not handling `TSAsExpression` types.

### LINKED ISSUE #306: [v4][no-debug] Crash with `Cannot read property 'name' of null`
We've got some code that looks like:

```js
thing.method.lastCall.args[0]();
```

This is causing a crash in the CallExpression handler in `no-debug`. In this case, `callExpressionIdentifier` is null. I suspect that `getDeepestIdentifierNode` doesn't handle array accessors correctly, and it also [can explicitly return `null`](https://github.com/testing-library/eslint-plugin-testing-library/blob/beta/lib/node-utils.ts#L433).

### PR DESCRIPTION
**Pull request by Belco90**
Closes #305. Closes #306.

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -6,12 +6,6 @@ module.exports = {
...
     global: {
-      branches: 100,
-      functions: 100,
-      lines: 100,
-      statements: 100,
-    },
-    './lib/node-utils.ts': {
       branches: 90,
       functions: 90,
       lines: 90,
```
Comment (Belco90) [inline jest.config.js:11]: I've dropped the coverage threshold a bit since I prefer to guard against null nodes everywhere even if there is no _possible_ code for it. We'll see what I do with this after releasing v4.

File: lib/rules/no-debug.ts [hunk 2]
```diff
@@ -83,13 +89,22 @@ export default createTestingLibraryRule<Options, MessageIds>({
       },
       CallExpression(node) {
         const callExpressionIdentifier = getDeepestIdentifierNode(node);
```
Comment (Belco90) [inline lib/rules/no-debug.ts:91]: @timdeschryver This is really weird: `getDeepestIdentifierNode` (and some other utils for finding identifiers) are typed as returning `TSESTree.Identifier | null`. However, TS is assuming they always returning `TSESTree.Identifier` and never `null`:

CleanShot 2021-04-02 at 12 02 28

Then, if I update the function to `return ''` instead of `return null` and update the returned type to `TSESTree.Identifier | string`, TS starts complaining about returned value saved in all these rules might not be an `Identifier`. Is this a bug on TS? Am I missing something else? I don't get why it always assumes `null` won't be returned.

File: tests/lib/rules/no-debug.test.ts [hunk 3]
```diff
@@ -139,6 +139,10 @@ ruleTester.run(RULE_NAME, rule, {
       debug()
       `,
     },
+
+    `// cover edge case for https://github.com/testing-library/eslint-plugin-testing-library/issues/306
+    thing.method.lastCall.args[0]();
```
Comment (Belco90) [inline tests/lib/rules/no-debug.test.ts:144]: Test for issue mentioned by @epmatsw

File: lib/rules/no-debug.ts (hunk 2 above)
Comment (epmatsw) [reply lib/rules/no-debug.ts:91]: I suspect turning on [strictNullChecks](https://www.typescriptlang.org/tsconfig#strictNullChecks) in the tsconfig would fix that.

Comment (timdeschryver) [reply lib/rules/no-debug.ts:91]: Oh wow... TIL...
@epmatsw is right, you'll have to turn the strict mode on.
I would encourage to turn on strict mode though, not just strict null checks.

Comment (Belco90) [reply lib/rules/no-debug.ts:91]: Aah damn! I thought we had `strict` enabled, but anyway I didn't expect those nullable values to be silent. I'm gonna enable strict mode and fix all corresponding issues then. Thanks guys.

**Review by timdeschryver — APPROVED** [review]

**Comment by Belco90** [comment]
Ok, so I've enabled TS `strict` mode and fix all issues. But now tests files generating invalid cases spreading an array of cases are flagged as errors by TS. I'm not sure what to do with that. Any ideas? @timdeschryver @epmatsw

**Comment by epmatsw** [comment]
Hmm. Maybe throw an `as const` somewhere in there?

**Comment by Belco90** [comment]
> Hmm. Maybe throw an `as const` somewhere in there?

That fixed almost all of them. The problem is the `messageId` for the errors, so TS complains about receiving an `string` but is declared as an enum. The `as const` does the trick for it. I'm learning more TS with this PR than in my whole life 😂 

Now I got some tests failing, I'm looking into them.

**Comment by Belco90** [comment]
Everything ok now, merging to test it on beta 3.
